pkg/operator/controller/enablednsnameresolver/controller.go

```diff
@@ -0,0 +1,131 @@
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &configv1.FeatureGate{}), &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DNSNameResolverEnabled indicates that the "DNSNameResolver" featuregate is enabled.
+	DNSNameResolverEnabled bool
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The enable_dnsnameresolver controller starts these caches when the
+	// "DNSNameResolver" featuregate is enabled. The dependent caches are dnsNameResolverCache
+	// and corednsEndpointsSliceCache.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The enable_dnsnameresolver controller starts these controllers when the
+	// "DNSNameResolver" featuregate is enabled and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual featuregate reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and starts the dependent
+// caches and controllers.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logrus.Info("reconciling", "request", request)
+
+	// Check if the DNSNameResolved feature gate is enabled or not.
+	if !r.config.DNSNameResolverEnabled {
+		return reconcile.Result{}, nil
+	}
+
+	// Start the dependent caches and wait for the caches to sync.
+	r.startCaches.Do(func() {
+		var wg sync.WaitGroup
+		for i := range r.config.DependentCaches {
+			cache := &r.config.DependentCaches[i]
+
+			// Start the dependent cache.
+			go func() {
+				if err := (*cache).Start(ctx); err != nil {
+					logrus.Errorf("failed to start cache, request.Name: %s, err: %v", request.Name, err)
+				}
+			}()
+
+			// Wait for the dependent cache to sync.
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if started := (*cache).WaitForCacheSync(ctx); !started {
+					logrus.Error("failed to sync cache before starting controllers")
+				}
+			}()
+		}
+		// Wait for all the dependent caches to sync.
+		wg.Wait()
+	})
+
+	// Start the dependent controllers after the dependent caches have synced.
+	r.startControllers.Do(func() {
+		for i := range r.config.DependentControllers {
+			controller := &r.config.DependentControllers[i]
+
+			// Start the dependent controller.
+			go func() {
+				if err := (*controller).Start(ctx); err != nil {
+					logrus.Errorf("failed to start controller, request.Name: %s, err: %v", request.Name, err)
+				}
+			}()
+		}
+	})
+
+	return reconcile.Result{}, nil
```

Comment:
If there are any errors logged here, for starting or syncing caches or controllers, no errors are returned.  I think that means it won't try to to do this again.  Is that planned?

Reply:
As mentioned in this discussion thread (https://github.com/openshift/cluster-dns-operator/pull/394#discussion_r1594874359), the enablednsnameresolver controller is planned to be removed in the future. After the removal the dependent controller and caches will be started by the manager itself, if the DNSNameResolver feature gate is enabled.

Reply:
This means that in the meantime, if there is any error starting or syncing caches or controllers, that it just remains broken without a status message or message for user?  Because even briefly, that wouldn't be acceptable.

Reply:
This same pattern is in use in cluster-ingress-operator for gateway API. I have followed the similar pattern. Her the differences are that the CRD is not being ensured and there are caches to be started in addition to the controller. https://github.com/openshift/cluster-ingress-operator/blob/8f98c618c5609cf7fcb97e5c61dc1e7a30576925/pkg/operator/controller/gatewayapi/controller.go#L91-L102

Reply:
@candita PLMK if you still need me to make any changes to the existing code or should we go ahead with the current code which is already an accepted pattern.

Reply:
Since this feature could start in Tech Preview, we should do a little more to let the user know if the feature failed to be enabled when they requested it.

Reply:
The reason an error is not returned is that even if we return an error, the starting of the caches and controller is tried only once. Thus returning an error will not have any effect. Additionally, we actually should  call the `Start` functions only once as they set up a lot of different resources. These functions are actually not retryable and will anyway throw error.

Reply:
The `Start` functions are blocking functions as well. As implemented currently, they need to be called in a separate goroutine and may block until the channel of the context being passed is closed. If we wait for the error to be returned by the `Start` functions then the flow will get completely blocked and CDO will become unresponsive to any cluster event.

Reply:
> If there are any errors logged here, for starting or syncing caches or controllers, no errors are returned.

This remained me of [the same concern](https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/2#discussion_r1575911553) I had with in the "standalone" operator from the plugin repository. Apart from this reporting issue we also risk to leave the main controller not started or without its caches.

> Additionally, we actually should call the Start functions only once as they set up a lot of different resources. These functions are actually not retryable and will anyway throw error.

Yeah, this may be a concern, I didn't check whether the cache's or controller's `Start` method is retryable. And yes we would need a stack of channels to report the errors and wait on these channels to retry.
This may be too much code to do. I would prefer to invest the same time into making the controller independent from any helper controller. I know we already decided to follow up on this later but I find that getting rid of the helper controller may easier (and more productive) than trying to render this code safer. I already have [a prototype PR](https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/4) which worked well at the time I submitted it.
@arkadeepsen: What if we review and test it tomorrow? Do you see a lot of work to be done in this PR once the main controller is independent?

Reply:
> @arkadeepsen: What if we review and test it tomorrow? Do you see a lot of work to be done in this PR once the main controller is independent?

The effort on this PR would be very minimal. It will just be changes to the operator.go file and removal of the enablednsnameresolver controller code.

Reply:
Thanks @alebedev87 for https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/4. I have pulled in the changes and removed the enablednsnameresolver controller.